src/condor_starter.V6.1/filter.cpp

```diff
@@ -0,0 +1,181 @@
+                } else {
+                    dprintf( D_NEVER, "illegal type %d for attribute %s\n", (int)v.GetType(), attribute.c_str() );
+
+                    std::string buffer;
+                    classad::ClassAdUnParser caup;
+                    caup.Unparse(buffer, e);
+                    dprintf( D_NEVER, "expression was '%s'\n", buffer.c_str() );
+                }
+        } else {
+            dprintf( D_NEVER, "failed to evaluate attribute %s\n", attribute.c_str() );
+
+            std::string buffer;
+            classad::ClassAdUnParser caup;
+            caup.Unparse(buffer, e);
+            dprintf( D_NEVER, "expression was '%s'\n", buffer.c_str() );
+        }
+    }
+
+    return filteredAd;
+}
+
+
+// The caller owns (is responsible for delete'ing) the returned ClassAd.
+classad::ClassAd *
+filterPluginResults( const classad::ClassAd & ad ) {
+    using namespace classad;
+
+    std::map<std::string, classad::Value::ValueType, CaseIgnLTStr> LegalAttributes {
+        { "TransferSuccess",        Value::BOOLEAN_VALUE },
+        { "TransferError",          Value::STRING_VALUE },
+        { "TransferProtocol",       Value::STRING_VALUE },
+        { "TransferType",           Value::STRING_VALUE },
+        { "TransferFileName",       Value::STRING_VALUE },
+        { "TransferFileBytes",      Value::INTEGER_VALUE },
+        { "TransferTotalBytes",     Value::INTEGER_VALUE },
+        { "TransferStartTime",      Value::INTEGER_VALUE },
+        { "TransferEndTime",        Value::INTEGER_VALUE },
+        { "ConnectionTimeSeconds",  Value::REAL_VALUE },
+        { "TransferURL",            Value::STRING_VALUE },
+        { "TransferErrorData",      Value::CLASSAD_VALUE },
+        { "DeveloperData",          Value::CLASSAD_VALUE },
+        { "TransferData",           Value::CLASSAD_VALUE },
+    };
+
+
+    // The ClassAd iterator isn't sorted, which set_intersection() requires.
+    classad::References attributes;
+    sGetAdAttrs( attributes, ad );
+
+    std::vector<std::string> intersection;
+#if defined(CXX_TWENTY_IS_REAL_TO_ME)
+    std::ranges::set_intersection(
+        attributes,
+        LegalAttributes | std::ranges::views::keys,
+        std::back_inserter( intersection ),
+        CaseIgnLTStr()
+    );
+#else
+    for( const auto & attribute : attributes ) {
+        if( LegalAttributes.count(attribute) != 0 ) {
+            intersection.push_back( attribute );
+        }
+    }
+#endif
+
+
+    if( intersection.empty() ) { return NULL; }
+
+    classad::ClassAd * filteredAd = new classad::ClassAd();
+    for( const auto & attribute : intersection ) {
+        classad::Value v;
+        ExprTree * e = ad.Lookup( attribute );
+        if( e->Evaluate(v) ) {
+                if( v.GetType() == LegalAttributes[attribute] ) {
+                    if( attribute == "TransferErrorData" ) {
+                        ClassAd * tedAd = NULL;
+                        ASSERT(v.IsClassAdValue(tedAd));
```

Comment:
do you really want to abort the process here if the value is not a classad?

Reply:
The conditional in line 151 assures us that the attribute named "TransferErrorData" is a ClassAd.  If we fail to extract a ClassAd from a classad::Value whose type is ClassAd, then yes, I think the process should abort.

Reply:
Fair enough.  only kill the process if there is a bug in the file transfer code.  Don't like it but I can live with it.